walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,128 @@
+      optionalEnd = checkIfEventConflicts(
+          optionalAvailableTimes, eventList.get(i), request, true, optionalEnd);
+    }
+    addFinalRange(mandatoryEnd, request, mandatoryAvailableTimes);
+    addFinalRange(optionalEnd, request, optionalAvailableTimes);
+
+    return (optionalAvailableTimes.isEmpty() && !request.getAttendees().isEmpty())
+        ? mandatoryAvailableTimes
+        : optionalAvailableTimes;
+  }
+
+  /**
+   * Comparator which organizes events based off the earliest start time
+   */
+  public int compare(Event a, Event b) {
+    // sort by start
+    return a.getWhen().start() - b.getWhen().start();
+  }
+
+  /**
+   * Checks if the set of event attendees has any members from the request
+   *
+   * @param eventAttendees the set of people attending the event
+   * @param request the specifications of the request used to retrieve attendees and optional
+   *     attendees
+   * @param checkingOptional a boolean if we are checking for optional attendees
+   * @return if any of the requested attendees exist in the list for eventAttendees
+   */
+  public static boolean hasAttendees(
+      Set eventAttendees, MeetingRequest request, boolean checkingOptional) {
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    for (String attendee : attendees) {
+      if (eventAttendees.contains(attendee)) {
+        return true;
+      }
+    }
+    if (checkingOptional) {
+      for (String optionalAttendee : optionalAttendees) {
+        if (eventAttendees.contains(optionalAttendee)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Checks if the time between the event and the last event is enough for the requested meeting
+   * time
+   *
+   * @param availableTimes a collection of times which will work for the request to be added to
+   * @param event the  event which we will check to see if there is enough time between its start
+   *     and the last
+   * meeting's end
+   * @param request the specifications for the requested meeting
+   * @param optional a boolean which determines if we are checking optional attendees as well
+   * @param start a number which represents the end of the last meeting
+   *
+   * @return the end of this event if applicable
+   */
+  public static int checkIfEventConflicts(Collection<TimeRange> availableTimes, Event event,
+      MeetingRequest request, boolean optional, int end) {
+    if (hasAttendees(event.getAttendees(), request, optional)) {
+      TimeRange eventRange = event.getWhen();
+
+      if (end < eventRange.start()) {
+        TimeRange range = TimeRange.fromStartEnd(end, eventRange.start(), false);
+        if (range.duration() >= request.getDuration()) {
+          availableTimes.add(range);
+        }
+      }
+      if (end < eventRange.end()) {
+        return eventRange.end();
+      }
```

Comment:
Can you add some comments explaining the logic here?

Reply:
Added